src/mlpack/core/tree/spill_tree/spill_tree_impl.hpp

```diff
@@ -0,0 +1,731 @@
+             class SplitType>
+inline size_t SpillTree<MetricType, StatisticType, MatType, HyperplaneType,
+    SplitType>::Descendant(const size_t index) const
+{
+  if (IsLeaf())
+    return (*pointsIndex)[index];
+  size_t num = left->NumDescendants();
+  if (index < num)
+    return left->Descendant(index);
+  if (right)
+    return right->Descendant(index - num);
+  // This should never happen.
+  return (size_t() - 1);
+}
+
+/**
+ * Return the index of a particular point contained in this node.
+ */
+template<typename MetricType,
+         typename StatisticType,
+         typename MatType,
+         template<typename HyperplaneMetricType> class HyperplaneType,
+         template<typename SplitMetricType, typename SplitMatType>
+             class SplitType>
+inline size_t SpillTree<MetricType, StatisticType, MatType, HyperplaneType,
+    SplitType>::Point(const size_t index) const
+{
+  if (IsLeaf())
+    return (*pointsIndex)[index];
+  // This should never happen.
+  return (size_t() - 1);
+}
+
+template<typename MetricType,
+         typename StatisticType,
+         typename MatType,
+         template<typename HyperplaneMetricType> class HyperplaneType,
+         template<typename SplitMetricType, typename SplitMatType>
+             class SplitType>
+inline bool SpillTree<MetricType, StatisticType, MatType, HyperplaneType,
+    SplitType>::HalfSpaceIntersects(const SpillTree& other) const
+{
+  if (!Parent())
+    return true;
+
+  const bool left = this == Parent()->Left();
+
+  if (left)
+    return !Parent()->Hyperplane().Right(other.Bound());
+  else
+    return !Parent()->Hyperplane().Left(other.Bound());
+}
+
+template<typename MetricType,
+         typename StatisticType,
+         typename MatType,
+         template<typename HyperplaneMetricType> class HyperplaneType,
+         template<typename SplitMetricType, typename SplitMatType>
+             class SplitType>
+template<typename VecType>
+inline bool SpillTree<MetricType, StatisticType, MatType, HyperplaneType,
+    SplitType>::HalfSpaceContains(
+    const VecType& point,
+    typename boost::enable_if<IsVector<VecType> >::type*) const
+{
+  if (!Parent())
+    return true;
+
+  const bool left = this == Parent()->Left();
+  const bool toTheLeft = Parent()->Hyperplane().Left(point);
+
+  return left == toTheLeft;
+}
+
+template<typename MetricType,
+         typename StatisticType,
+         typename MatType,
+         template<typename HyperplaneMetricType> class HyperplaneType,
+         template<typename SplitMetricType, typename SplitMatType>
+             class SplitType>
+void SpillTree<MetricType, StatisticType, MatType, HyperplaneType, SplitType>::
+    SplitNode(std::vector<size_t>& points,
+              const size_t maxLeafSize,
+              const double tau,
+              const double rho)
+{
+  // We need to expand the bounds of this node properly.
+  for (size_t i = 0; i < points.size(); i++)
+    bound |= dataset->cols(points[i], points[i]);
```

Comment:
I have simplified it in: https://github.com/mlpack/mlpack/pull/747/commits/071588457415094e99883774af541075e3858c22